subsys/bluetooth/mesh/key_importer.c

```diff
@@ -0,0 +1,420 @@
+	node.num_elem = legacy_node.num_elem;
+	node.flags = legacy_node.flags;
+	memcpy(node.uuid, legacy_node.uuid, 16);
+
+	err = settings_save_one(path, &node, sizeof(node));
+	if (err) {
+		LOG_ERR("Failed to store cdb node: %s (err %d)", path, err);
+		return err;
+	}
+
+	return 0;
+}
+
+static int get_idx(const char *name)
+{
+	const char *next;
+
+	settings_name_next(name, &next);
+
+	if (!next) {
+		return -EINVAL;
+	}
+
+	return strtol(next, NULL, 16);
+}
+
+static int load_keys_cb(const char *name, size_t len, settings_read_cb read_cb, void *cb_arg,
+			void *param)
+{
+	const char *next;
+	char path[28];
+	int idx;
+
+	if (!name) {
+		LOG_ERR("Insufficient number of arguments");
+		return -ENOENT;
+	}
+
+	LOG_DBG("Load name: %s", name);
+
+	if (!strncmp(name, "NetKey", strlen("NetKey"))) {
+		idx = get_idx(name);
+
+		if (idx < 0) {
+			goto not_found;
+		}
+
+		snprintk(path, sizeof(path), "bt/mesh/NetKey/%x", idx);
+		return import_net_keys(path, len, read_cb, cb_arg);
+	}
+
+	if (!strncmp(name, "AppKey", strlen("AppKey"))) {
+		idx = get_idx(name);
+
+		if (idx < 0) {
+			goto not_found;
+		}
+		snprintk(path, sizeof(path), "bt/mesh/AppKey/%x", idx);
+		return import_app_keys(path, len, read_cb, cb_arg);
+	}
+
+	if (!strncmp(name, "Net", strlen("Net"))) {
+		snprintk(path, sizeof(path), "bt/mesh/Net");
+		return import_dev_keys(path, len, read_cb, cb_arg);
+	}
+
+	if (!strncmp(name, "DevKeyC", strlen("DevKeyC"))) {
+		snprintk(path, sizeof(path), "bt/mesh/DevKeyC");
+		return import_devcand_keys(path, len, read_cb, cb_arg);
+	}
+
+	if (strncmp(name, "cdb", strlen("cdb"))) {
+		return 0;
+	}
+
+	settings_name_next(name, &next);
+
+	if (!next) {
+		goto not_found;
+	}
+
+	if (strncmp(next, "Net", strlen("Net")) == 0) {
+		return 0;
+	}
+
+	idx = get_idx(next);
+
+	if (idx < 0) {
+		goto not_found;
+	}
+
+	if (!strncmp(next, "Node", strlen("Node"))) {
+		snprintk(path, sizeof(path), "bt/mesh/cdb/Node/%x", idx);
+		return import_cdb_dev_keys(path, len, read_cb, cb_arg);
+	}
+
+	if (!strncmp(next, "Subnet", strlen("Subnet"))) {
+		snprintk(path, sizeof(path), "bt/mesh/cdb/Subnet/%x", idx);
+		return import_net_keys(path, len, read_cb, cb_arg);
+	}
+
+	if (!strncmp(next, "AppKey", strlen("AppKey"))) {
+		snprintk(path, sizeof(path), "bt/mesh/cdb/AppKey/%x", idx);
+		return import_app_keys(path, len, read_cb, cb_arg);
+	}
+
+	return 0;
+
+not_found:
+	LOG_ERR("Unknown key %s", name);
+	return -ENOENT;
+}
+
+static int import_keys(void)
+{
+	int err;
+
+	err = settings_subsys_init();
+	if (err) {
+		LOG_ERR("Settings initialization failed (err %d)", err);
+		return err;
+	}
+
+	err = bt_mesh_crypto_init();
+	if (err) {
+		LOG_ERR("Crypto librarry initialization failed (err %d)", err);
+		return err;
+	}
+
+	LOG_DBG("Importing keys...");
+	settings_load_subtree_direct("bt/mesh", load_keys_cb, NULL);
+	LOG_DBG("Importing keys done");
+
+	return 0;
+}
+
+SYS_INIT(import_keys, APPLICATION, CONFIG_KERNEL_INIT_PRIORITY_DEFAULT);
```

Comment:
Should we make priority configurable? It can default to `KERNEL_INIT_PRIORITY_DEFAUL`. Unlikely this will have conflicts, but still.